File: JavaBankSystem/src/service/AuthService.java
package service;

import java.util.HashMap;

import data.DataStore;
import model.User;
import util.PasswordUtil;

public class AuthService {
    private HashMap<String, User> users;

    public AuthService() {
        this.users = DataStore.getUsers();
    }

    public boolean register(String username, String password) {
        if(DataStore.existsUser(username)) {
            System.out.println("Username already exists. Please choose a different username.");
            return false;
        }

        if(!PasswordUtil.isValidPassword(password)) {
            System.out.println("Password must be at least 8 characters long and include uppercase, lowercase, number, and special character.");
            return false;
        }

        String hash = PasswordUtil.hashPassword(password);

        User user = new User(username, hash);

        users.put(hash, user);

        System.out.println("User registered successfully.");
        return true;
    }

    public User login(String username, String password) {
        String hash = PasswordUtil.hashPassword(password);

        User user = users.get(hash);

        if(user == null) {
            System.out.println("Invalid username or password.");
            return null;
        }

        if(!user.getPasswordHash().equals(hash)) {
            System.out.println("Invalid username or password.");
            return null;
        }

        System.out.println("Login successful. Welcome, " + user.getUsername() + "!");
        return user;
    }
}

File: CppBankSystem/src/service/AuthService.cpp
#include "AuthService.h"
#include "../data/DataStore.h"
#include "../util/PasswordUtil.h"
#include <iostream>

AuthService::AuthService() : users(DataStore::getUsers()) {}

bool AuthService::registerUser(const std::string& username, const std::string& password) {
    if(DataStore::existsUser(username)) {
        std::cout << "Username already exists. Please choose another one." << std::endl;
        return false;
    }

    if(!PasswordUtil::isValidPassword(password)) {
        std::cout << "Password must be at least 8 characters long and include uppercase, lowercase, number, and special character." << std::endl;
        return false;
    }

    std::string hash = PasswordUtil::hashPassword(password);

    User user(username, hash);

    auto result = users.emplace(username, User(username, hash));

    if (!result.second) {
        std::cout << "User already exists." << std::endl;
    }

    std::cout << "User registered successfully!" << std::endl;
    return true;
}

User* AuthService::loginUser(const std::string& username, const std::string& password) {

    auto it = users.find(username);

    if (it == users.end()) {
        std::cout << "Invalid username or password." << std::endl;
        return nullptr;
    }

    std::string hash = PasswordUtil::hashPassword(password);

    User& user = it->second;

    if (user.getPasswordHash() != hash) {
        std::cout << "Invalid username or password." << std::endl;
        return nullptr;
    }

    std::cout << "Login successful. Welcome, " << user.getUsername() << "!" << std::endl;

    return &user;
}